Selection Modification

ericmehl opened the pull request:
This adds a mechanism for modifying what scene locations are actually selected when making a viewport selection. Users can register functions in Python or C++ that get a scene and the selection made by the user, and return the selection they want as a result.

It also includes a first use of the mechanism in the form of USD Kind selection. Starting at the viewport selected location, the first ancestor, or the original location itself, that is the USD Kind chosen in the `selectionModifier` plug is what is actually selected.

I tried a few different methods of presenting the USD Kinds in the `selectionModifier` widget. USD Kinds are hierarchical (https://openusd.org/dev/api/kind_page_front.html#mainpage_kind) so, for example, setting the `selectionModifier` to select USD Kind of `model` will also select `component`, `group` and `assembly`. Currently I'm expressing that by spelling out all the Kinds that will be considered a match in the menu item.

That was originally driven by using the standard presets metadata and dropdown. I think a nicer widget is what I have now, very similar to the `displayTransform.name` widget. Before settling on this presentation, I also looked at indenting based on the hierarchy, and even a little ASCII art to draw out elbows and lines to form a tree look.

Indenting with the default presets widget meant that when you make your selection, the label in the collapsed dropdown is indented too, which felt rather odd.

Now that I'm using a custom widget anyways, maybe the indented presentation is worth another look?

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

ericmehl commented:
I made the changes we talked about yesterday in the team call, namely :
- Using just "Select" as the label for the plug in the UI. That seemed a little unclear to use by itself for the actual plug name, so I went with `selectModePlug` and similar `selectMode*` throughout.
- Along with that, the default is now called "Any".
- Use `attributes()` instead of `fullAttributes()` so we get the selection to the location that actually holds the attribute we're looking for, instead of one that's inheriting it.
- Remove the verbose mode labeling and go with "Assembly", "Group", etc.

I'm looking at the jumpy UI when switching from Select to Translate and other tools that add additional UI elements alongside this new `selectMode` plug.

ericmehl commented:
I found a couple of ways to fix the jump in the UI. One is to add a 4px `margin-top` and `margin-bottom` to the style sheet, targeted at the widget for `SelectionToolUI.SelectModePlugValueWidget`.

What I went with instead is adding a `GafferUI.Frame` around the `SelectionToolUI.SelectModePlugValueWidget.__menuButton`. That works just as well, and is less complicated because it doesn't modify the style sheet. Curiously, this solution needs as little as a 1px `borderWidth` for the frame wrapper. I would have expected it to need 4px to match the `TransformTool._SselectionWidget` frame.

I'm open to other solutions if something seems better.

Otherwise, ready for review!

johnhaddon commented on src/GafferSceneUIModule/ToolBinding.cpp:
```diff
@@ -183,12 +183,57 @@ object acquireTransformEdit( const TransformTool::Selection &s, bool createIfNec
+		}
+	}
+	private :
+		object m_fn;
```
I think we'll get away with this because it's unlikely that anything is registered or deregistered from pure C++, but I'll mention it anyway : the destructor for `SelectModeHelper` will destroy this Python object without acquiring the GIL first. That has caused crashes in registries that are modified from pure C++. An alternative approach is the one I took in the Renderer registry recently : https://github.com/GafferHQ/gaffer/blob/main/src/GafferSceneModule/RenderBinding.cpp#L121-L142. The lambda there also cuts down on the amount of boilerplate a little bit. Might be worth considering...

ericmehl replied:
That looks cleaner to me, and worth fixing before it becomes a surprise problem later on. Fixed in https://github.com/GafferHQ/gaffer/pull/5720/commits/342035f8dfdc6dc7d5b683e025d388f68f12a6a3

johnhaddon commented on Changes.md:
```diff
@@ -26,6 +26,11 @@ Documentation
+Features
+--------
+
+- SelectionTool : Added selection modifier plug. When set to anything except `None` using the SelectionTool causes the actual scene location selected to potentially be modified from the originally selected location. Selection modifiers work identically for deselection. Currently, USD Kind modifiers are implemented. When selecting, the first ancestor location with a `usd:kind` attribute matching t ...
```
"selection modifier" -> "select mode". "None" -> "Any".

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5720/commits/13cb3fbfe4bf47f4269ecb9b3598bc97b152aead, with "None" -> "Standard" to reflect that separate change.

johnhaddon commented on startup/gui/usd.py:
```diff
@@ -45,3 +53,48 @@
+
+	kind = None
+	while len( path ) > 0 :
+		attributes = scene.attributes( IECore.InternedStringVectorData( path ) )
```
Can we omit the `InternedStringVectorData` here, since Murray added a conversion from lists recently?

ericmehl replied:
Very nice, fixed in https://github.com/GafferHQ/gaffer/pull/5720/commits/cde068adae72c742305aab275801a2973a37e267

johnhaddon commented on startup/gui/usd.py:
```diff
@@ -45,3 +53,48 @@
+			break
+		path.pop()
+
+	return "/" + "/".join( path )
```
Likewise, I think here we might just be able to return `path` now?

ericmehl replied:
Same as above, fixed in https://github.com/GafferHQ/gaffer/pull/5720/commits/cde068adae72c742305aab275801a2973a37e267

johnhaddon commented on python/GafferSceneUI/SelectionToolUI.py:
```diff
@@ -112,3 +106,78 @@ class _RightSpacer( GafferUI.Spacer ) :
+
+			if category != "Any" and category not in modifierCategories :
+				result.append( f"/__{category}Dividier", { "divider" : True, "label" : category } )
+				modifierCategories.add( category )
```
Probably unimportant, but there's an assumption here that all the things in one category will be grouped together in `registeredSelectModes()` - if not, then things end up under the wrong category. Probably fixable using `insertAfter()`, and using `modifiedCategories` to map to the last item in each category.

Like I say, unimportant. The only reason I mention is that I quite like this presentation, so I'm musing about it being an optional mode for PresetsPlugValueWidget, or even Menu. For that we'd need to be more robust I think.

ericmehl replied:
May as well fix it now before someone gets frustrated trying to do something clever with the menu. I think I have a good solution in https://github.com/GafferHQ/gaffer/pull/5720/commits/00a56aa5e1225494185a1219867516da4ac76664

johnhaddon commented on python/GafferSceneUI/SelectionToolUI.py:
```diff
@@ -111,9 +111,10 @@ class SelectModePlugValueWidget( GafferUI.PlugValueWidget ) :
 	def __init__( self, plugs, **kw ) :
 
-		self.__menuButton = GafferUI.MenuButton( "", menu = GafferUI.Menu( Gaffer.WeakMethod( self.__menuDefinition ) ) )
+		with GafferUI.Frame( borderWidth = 1, borderStyle = GafferUI.Frame.BorderStyle.None_, **kw ) as outerFrame :
```
It seems wrong to be doing this here, in a widget that shouldn't really care about the layout it is in. One alternative might be to give `_LeftSpacer` a height to match the highest thing. That seems more appropriate since the spacer is more about fiddling the details of the layout as a whole. Did you try reducing the size of the "Editing..." frame instead? Maybe that looks a bit rubbish?

ericmehl replied:
Maybe not quite rubbish, but the extra breathing space around the selection tip does feel nice to me.

Here's the selection tip (gray box) with border = 4 as it is now :
![image](https://github.com/GafferHQ/gaffer/assets/7873467/10181e37-7590-4c0b-9473-5704577af4c1)

and border = 1 (largest border that doesn't cause the jump) :

![image](https://github.com/GafferHQ/gaffer/assets/7873467/048e252b-76d4-4f3a-8565-399cd05b0837)

I added some height to the `LeftSpacer`, that does seem clearer to keep the details in that widget instead of `SelectModePlugValueWidget`. The 27 pixel value I got from trial and error.

johnhaddon reviewed (commented):
Thanks Eric, and once again, sorry for the delay in getting to the review. This is a nice addition, and I like how you're querying the kind registry to work with anything in there, rather than just hardcoding it.

In addition to the inline comments, here's a couple of other talking points :

- I think we discussed synchronising the select mode across all the tools? Currently they each have their own mode, which is going to get pretty confusing I think.
- I don't have a strong rationale either way, but I do wonder if "Standard" might be better than "Any" as the default mode. Or maybe "Default"?
- I wonder if there are any non-usd modes we could provide out of the box? First-ancestor-with-a-shader-assignment perhaps? Not that I can think of a concise label for that right now.

Cheers...
John

ericmehl commented on include/GafferSceneUI/SelectionTool.h:
```diff
@@ -64,6 +66,26 @@ class GAFFERSCENEUI_API SelectionTool : public GafferUI::Tool
+		static void registerSelectMode(
+			const std::string &name,
+			SelectFunction function,
+			const std::string label = ""
```
It's perhaps a bit awkward to make the menu setup work, but not too bad. In https://github.com/GafferHQ/gaffer/pull/5720/commits/e744857485fed7d3531d31ce5cf6ab465860dd55 I removed the label field and have the popup menu partition on `/` to get the first category name. `Any` (soon to be `Standard`) I prefixed with `/` so it gets an empty string as the partition prefix and it now works without special handling for the menu build.

ericmehl commented on src/GafferSceneUI/SelectionTool.cpp:
```diff
@@ -54,6 +58,71 @@ using namespace GafferUI;
+	// shut down first).
+	static auto g_selectModes = new SelectModeMap;
+
+	if( ( *g_selectModes ).empty() )
```
I prefer that too. This formatting must have suck in modeling it after existing code maybe. Fixed in https://github.com/GafferHQ/gaffer/pull/5720/commits/4fbc2eff973e0d2e77c54480a22ff41dade78a29

ericmehl commented on src/GafferSceneUI/SelectionTool.cpp:
```diff
@@ -227,6 +336,12 @@ bool SelectionTool::buttonPress( const GafferUI::ButtonEvent &event )
 	ScenePlug::ScenePath objectUnderMouse;
 	sg->objectAt( event.line, objectUnderMouse );
 
+	objectUnderMouse = modifyPath(
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5720/commits/7174ef97d3a705bc0a691b7e9671c7659e2cddaf

ericmehl commented:
> * I think we discussed synchronising the select mode across all the tools? Currently they each have their own mode, which is going to get pretty confusing I think.

Right! That had slipped my mind. I'll look at that now that I have the other feedback addressed.



> * I don't have a strong rationale either way, but I do wonder if "Standard" might be better than "Any" as the default mode. Or maybe "Default"?

"Standard" seems pretty good to me, it sounds a bit more confident than "Default" to my mind. That is fixed over the course of https://github.com/GafferHQ/gaffer/pull/5720/commits/29cb09cceef2893c142c7cac4cc5af6853e27528, https://github.com/GafferHQ/gaffer/pull/5720/commits/95c72d58c153d794595620493c95e9683146ebb3 and https://github.com/GafferHQ/gaffer/pull/5720/commits/cc5877d3facc6a7b52ddbf6951746e63c78ca947 for the purposes of squishability.

> * I wonder if there are any non-usd modes we could provide out of the box? First-ancestor-with-a-shader-assignment perhaps? Not that I can think of a concise label for that right now.

That sounds like a nice one to have too, I'll add that in with something better than First-ancestor-with-a-shader-assignment or, my working title, "Shader-Bequether"

ericmehl commented:
I added selection synchronization across the tools in https://github.com/GafferHQ/gaffer/pull/5720/commits/dc7c007696eafa292f79dff34defa378602159eb. And in https://github.com/GafferHQ/gaffer/pull/5720/commits/65ea3dad6782357921c538407cc6deaba6c6655a, I added a mode to select the ancestor (or current location) that has either a surface or displacement shader.

Ready for a new look.

johnhaddon commented:
> Ready for a new look.

Thanks Eric, looks great! I like the new shader-bequeather modes :) Two small things :

- Could we capitalize the names for the USD modes please? (in the registration code, not the menu-building code).
- Should we target this to `main` since that is where https://github.com/GafferHQ/gaffer/pull/5737? We're really close to getting 1.4 out the door, and I don't think this is urgent enough that it needs to go in 1.3.x.

ericmehl commented:
> * Could we capitalize the names for the USD modes please? (in the registration code, not the menu-building code).

USD Kind comparisons are case-sensitive, so I didn't do this initially in order to not make assumptions when converting to and from the native Kind name. Do you think `IECore.CamelCase.toSpaced()` and `.fromSpaced()` would be appropriate? Or just assuming all USD Kinds, including custom Kinds a user might register, are lower case?

> * Should we target this to `main`

Yeah, with #5737 is on `main` and 1.4 is so close, that sounds good to me.

johnhaddon commented:
> Do you think IECore.CamelCase.toSpaced() and .fromSpaced() would be appropriate?

I think `toSpaced()` might be a reasonable way of formatting the first argument to `registerMode()`. But I don't know why we would need `fromSpaced()` at all. Oh, I see, you're deriving the original kind from the formatted name in `__kindSelectionModifier` - we don't need to do that. We should just pass the original true kind name to that...

ericmehl commented:
> We should just pass the original true kind name to that...

Oh yes, of course. Had a brain blockage on that one.

johnhaddon commented:
Cool - if you could get this all squashed down and ready to merge for `main` that would be awesome. Thanks!

johnhaddon commented:
Thanks Eric!
